### DISCUSSION TITLE
macOS: usb device interface is undefined.

### DISCUSSION THREAD

**Discussion by emAsif** [libusb/libusb#1529] [Q&A]
After installing usb library in project, when I try to access the usb device it shows all the information except for device interface which is undefined. I think the device is bound to kernal and since its an HID device I cannot deattach it using root privilege but still the device interface being undefined is strange. I'm new to mac kernal interaction and did some online research and found that this is well known issue with mac usb connection. I found a resource which talks about IOKit extension [https://developer.apple.com/library/archive/documentation/DeviceDrivers/Conceptual/USBBook/USBDeviceInterfaces/USBDevInterfaces.html#//apple_ref/doc/uid/TP40002645-TPXREF101](https://github.com/node-usb/node-usb/issues/url) and there are some other options like sudo kextunload which is not long term solution and also didn't work for my mac.

Additional Information
device type: usb-2.0-tracking-device
Operating System: macOS Big Sonoma
Runtime (node -v: 18.2.2 , electron -v: 31.1.0,):
node-usb: (2.13.0):

**Comment by mcuee** [comment]
Please take note that node-usb is not under libusb. Please raise the issue there.

**Comment by mcuee** [comment]
Reference: issue already raised under node-usb.
* https://github.com/node-usb/node-usb/issues/786

**Comment by mcuee** [comment]
For USB HID project, the recommendation is to use HIDAPI instead.

Reference:
https://github.com/libusb/libusb/wiki/FAQ#user-content-Does_libusb_support_USB_HID_devices

**Comment by emAsif** [comment]
@mcuee  I tried using HIDAPI but its not detected as HID device also I need to access the interface and cofigurations which seems not possible with USB HID project. I tried using a libusb instead of node-usb using a wrapper but still the issue persists. I found this system log error when i connect my device. can you please help with this issue. 

`Electron@(null): AppleUSBHostUserClient::openGated: failed to open GPS Plus Collar@00120000: provider is already opened for exclusive access by pid 11674, Microsoft Teams`

**Comment by mcuee** [comment]
Sorry I can not help you much here. But I will convert this to a discussion and hopefully others can help you better.

But forget about using libusb for an HID device under macOS. It is not supported and it will not work.

You can try to post the issue to hidapi project -- but please do not use a wrapper, use pure C when you want to report an issue.
